src/main/java/org/wso2/diagnostics/watchers/logwatcher/LogWatcher.java

```diff
@@ -120,31 +166,95 @@ private void closeQuietly(Closeable closeable) {
         }
     }
 
-    private void readLines(RandomAccessFile reader, StringBuilder logBuilder) throws IOException {
+    /**
+     * Class to hold processing state results
+     */
+    private static class ProcessingResult {
+        ProcessingState state;
+        String errorLine;
+        long errorTimestamp;
+        int postErrorLogLevelCount;
+
+        ProcessingResult(ProcessingState state, String errorLine, long errorTimestamp, int postErrorLogLevelCount) {
+            this.state = state;
+            this.errorLine = errorLine;
+            this.errorTimestamp = errorTimestamp;
+            this.postErrorLogLevelCount = postErrorLogLevelCount;
+        }
+    }
+
+    private ProcessingResult readLines(RandomAccessFile reader, LinkedList<String> contextQueue,
+                                       ProcessingState currentState, String errorLine,
+                                       long errorTimestamp, int postErrorLogLevelCount) throws IOException {
 
-        // Read the new lines
         String line;
-        String errorLine = "";
+
         while ((line = reader.readLine()) != null) {
-            // Check if the line indicates the start of a stack trace
-            if (line.contains("ERROR") || line.contains("WARN")) {
-                if (logBuilder.length() == 0) {
-                    errorLine = line;
-                    logBuilder.append(line).append("\n");
-                } else {
-                    interpreter.interpret(errorLine, logBuilder.toString());
-                    logBuilder.setLength(0);
-                    errorLine = line;
-                    logBuilder.append(line).append("\n");
-                }
-            } else {
-                logBuilder.append(line).append("\n");
-            }
-            // If the logBuilder exceeds a threshold, interpret the log lines and clear the logBuilder
-            if (logBuilder.length() > 1000000) {
-                interpreter.interpret(errorLine, logBuilder.toString());
-                logBuilder.setLength(0);
+            boolean isLogLevelLine = LOG_LEVEL_PATTERN.matcher(line).find();
+
+            switch (currentState) {
+                case NORMAL:
+                    if (isLogLevelLine) {
+                        // Manage pre-error context queue
+                        if (contextQueue.size() >= MAX_CONTEXT_SIZE) {
+                            contextQueue.removeFirst();
+                        }
+                        contextQueue.add(line);
+
+                        // Check if it's an error line
+                        if (line.contains("ERROR")) {
+                            currentState = ProcessingState.ERROR_COLLECTING;
+                            errorLine = line;
+                            errorTimestamp = System.currentTimeMillis();
+                        }
+                    } else {
+                        // Non-log level line in normal state, likely a stacktrace
+                        contextQueue.add(line);
```

Comment:
Shall we check logging multiple-line INFO logs and see whether this logic still works? For eg: using a log mediator which prints the full log.